service/matching/taskReader.go

```diff
@@ -421,3 +376,120 @@ func (tr *taskReader) newDispatchContext(isolationGroup string) (context.Context
 	}
 	return tr.cancelCtx, func() {}
 }
+
+func (tr *taskReader) dispatchSingleTaskFromBufferWithRetries(isolationGroup string, taskInfo *persistence.TaskInfo) (breakDispatchLoop bool) {
+	// retry loop for dispatching a single task
+	for {
+		breakDispatchLoop, breakRetryLoop := tr.dispatchSingleTaskFromBuffer(isolationGroup, taskInfo)
+		if breakRetryLoop {
+			return breakDispatchLoop
+		}
+	}
+}
+
+func (tr *taskReader) dispatchSingleTaskFromBuffer(isolationGroup string, taskInfo *persistence.TaskInfo) (breakDispatchLoop bool, breakRetries bool) {
+	task := newInternalTask(taskInfo, tr.completeTask, types.TaskSourceDbBacklog, "", false, nil, isolationGroup)
+	dispatchCtx, cancel := tr.newDispatchContext(isolationGroup)
+	timerScope := tr.scope.StartTimer(metrics.AsyncMatchLatencyPerTaskList)
+	err := tr.dispatchTask(dispatchCtx, task)
+	timerScope.Stop()
+	cancel()
+	if err == nil {
+		return false, true
+	}
+	if err == context.Canceled {
+		tr.logger.Info("Tasklist manager context is cancelled, shutting down")
+		return true, true
+	}
+	if err == context.DeadlineExceeded {
+		// it only happens when isolation is enabled and there is no pollers from the given isolation group
+		// if this happens, we don't want to block the task dispatching, because there might be pollers from
+		// other isolation groups, we just simply continue and dispatch the task to a new isolation group which
+		// has pollers
+		tr.logger.Warn("Async task dispatch timed out", tag.IsolationGroup(isolationGroup))
+		tr.scope.IncCounter(metrics.AsyncMatchDispatchTimeoutCounterPerTaskList)
+
+		// the idea here is that by re-fetching the isolation-groups, if something has shifted
+		// it will get a new isolation group to be placed. If it needs re-routing, then
+		// this will be the new routing destination.
+		group, err := tr.getIsolationGroupForTask(tr.cancelCtx, taskInfo)
+		if err != nil {
+			// it only errors when the tasklist is a sticky tasklist and
+			// the sticky pollers are not available, in this case, we just complete the task
+			// and let the decision get timed out and rescheduled to non-sticky tasklist
+			if err == _stickyPollerUnavailableError {
+				tr.completeTask(taskInfo, nil)
+				return false, true
+			}
+			// it should never happen, unless there is a bug in 'getIsolationGroupForTask' method
+			tr.logger.Error("taskReader: unexpected error getting isolation group", tag.Error(err))
+			tr.completeTask(taskInfo, err)
+			return false, true
+		}
+
+		if group == isolationGroup {
+			// no change, retry to dispatch the task again
+			return false, false
+		}
+
+		// ensure the isolation group is configured and available
+		_, taskGroupReaderIsPresent := tr.taskBuffers[group]
+		if !taskGroupReaderIsPresent {
+			// there's a programmatic error. Something has gone wrong with tasklist instantiation
+			// don't block and redirect to the default group
+			tr.scope.IncCounter(metrics.BufferIsolationGroupRedirectFailureCounter)
```

Comment:
do you think it would be useful to also include `isolationGroup` in the error message? same applies to other log messages below

Reply:
I thought about it, but decided against it because we don't have a tag specifically for this kind of key (we're pretty disciplined about which log tags get used) and it can be derived by looking up the task with the taskID or the workflow + runID

Reply:
Got it if we have a workaround and we think this will be super rare troubleshooting scenario no need to have that extra info as a log tag. Still could be baked into the message if you think it would be useful. 
`tr.logger.Error("Isolation group %s is missing task buffer. Task originally was assigned to group %s. Redirecting to default", tags...)`

Reply:
after debugging I completely agree, let me address

Reply:
added with some tag.Dynamic tags. I was not wanting to template the log message just because it can make searching tricky